File: src/CalcManager/ExpressionCommand.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

#include <string>
#include "Header Files/CCommand.h"
#include "ExpressionCommand.h"

using namespace std;
using namespace CalcEngine;

constexpr wchar_t chNegate = L'-';
constexpr wchar_t chExp = L'e';
constexpr wchar_t chPlus = L'+';

CParentheses::CParentheses(_In_ int command)
    : m_command(command)
{
}

int CParentheses::GetCommand() const
{
    return m_command;
}

CalculationManager::CommandType CParentheses::GetCommandType() const
{
    return CalculationManager::CommandType::Parentheses;
}

void CParentheses::Accept(_In_ ISerializeCommandVisitor& commandVisitor)
{
    commandVisitor.Visit(*this);
}

CUnaryCommand::CUnaryCommand(int command)
{
    m_command = make_shared<vector<int>>();
    m_command->push_back(command);
}

CUnaryCommand::CUnaryCommand(int command1, int command2)
{
    m_command = make_shared<vector<int>>();
    m_command->push_back(command1);
    m_command->push_back(command2);
}

const shared_ptr<vector<int>>& CUnaryCommand::GetCommands() const
{
    return m_command;
}

CalculationManager::CommandType CUnaryCommand::GetCommandType() const
{
    return CalculationManager::CommandType::UnaryCommand;
}

void CUnaryCommand::SetCommand(int command)
{
    m_command->clear();
    m_command->push_back(command);
}

void CUnaryCommand::SetCommands(int command1, int command2)
{
    m_command->clear();
    m_command->push_back(command1);
    m_command->push_back(command2);
}

void CUnaryCommand::Accept(_In_ ISerializeCommandVisitor& commandVisitor)
{
    commandVisitor.Visit(*this);
}

CBinaryCommand::CBinaryCommand(int command)
    : m_command(command)
{
}

void CBinaryCommand::SetCommand(int command)
{
    m_command = command;
}

int CBinaryCommand::GetCommand() const
{
    return m_command;
}

CalculationManager::CommandType CBinaryCommand::GetCommandType() const
{
    return CalculationManager::CommandType::BinaryCommand;
}

void CBinaryCommand::Accept(_In_ ISerializeCommandVisitor& commandVisitor)
{
    commandVisitor.Visit(*this);
}

COpndCommand::COpndCommand(shared_ptr<vector<int>> const& commands, bool fNegative, bool fDecimal, bool fSciFmt)
    : m_commands(commands)
    , m_fNegative(fNegative)
    , m_fSciFmt(fSciFmt)
    , m_fDecimal(fDecimal)
    , m_fInitialized(false)
    , m_value{}
{
}

void COpndCommand::Initialize(Rational const& rat)
{
    m_value = rat;
    m_fInitialized = true;
}

const shared_ptr<vector<int>>& COpndCommand::GetCommands() const
{
    return m_commands;
}

void COpndCommand::SetCommands(shared_ptr<vector<int>> const& commands)
{
    m_commands = commands;
}

void COpndCommand::AppendCommand(int command)
{
    if (m_fSciFmt)
    {
        ClearAllAndAppendCommand(static_cast<CalculationManager::Command>(command));
    }
    else
    {
        m_commands->push_back(command);
    }

    if (command == IDC_PNT)
    {
        m_fDecimal = true;
    }
}

void COpndCommand::ToggleSign()
{
    for (int nOpCode : *m_commands)
    {
        if (nOpCode != IDC_0)
        {
            m_fNegative = !m_fNegative;
            break;
        }
    }
}

void COpndCommand::RemoveFromEnd()
{
    if (m_fSciFmt)
    {
        ClearAllAndAppendCommand(CalculationManager::Command::Command0);
    }
    else
    {
        const size_t nCommands = m_commands->size();

        if (nCommands == 1)
        {
            ClearAllAndAppendCommand(CalculationManager::Command::Command0);
        }
        else
        {
            int nOpCode = m_commands->at(nCommands - 1);

            if (nOpCode == IDC_PNT)
            {
                m_fDecimal = false;
            }

            m_commands->pop_back();
        }
    }
}

bool COpndCommand::IsNegative() const
{
    return m_fNegative;
}

bool COpndCommand::IsSciFmt() const
{
    return m_fSciFmt;
}

bool COpndCommand::IsDecimalPresent() const
{
    return m_fDecimal;
}

CalculationManager::CommandType COpndCommand::GetCommandType() const
{
    return CalculationManager::CommandType::OperandCommand;
}

void COpndCommand::ClearAllAndAppendCommand(CalculationManager::Command command)
{
    m_commands->clear();
    m_commands->push_back(static_cast<int>(command));
    m_fSciFmt = false;
    m_fNegative = false;
    m_fDecimal = false;
}

const wstring& COpndCommand::GetToken(wchar_t decimalSymbol)
{
    static const wchar_t chZero = L'0';

    const size_t nCommands = m_commands->size();
    m_token.clear();

    for (size_t i = 0; i < nCommands; i++)
    {
        int nOpCode = (*m_commands)[i];

        if (nOpCode == IDC_PNT)
        {
            m_token += decimalSymbol;
        }
        else if (nOpCode == IDC_EXP)
        {
            m_token += chExp;
            int nextOpCode = m_commands->at(i + 1);
            if (nextOpCode != IDC_SIGN)
            {
                m_token += chPlus;
            }
        }
        else if (nOpCode == IDC_SIGN)
        {
            m_token += chNegate;
        }
        else
        {
            wstring num = to_wstring(nOpCode - IDC_0);
            m_token.append(num);
        }
    }

    // Remove zeros
    for (size_t i = 0; i < m_token.size(); i++)
    {
        if (m_token.at(i) != chZero)
        {
            if (m_token.at(i) == decimalSymbol)
            {
                m_token.erase(0, i - 1);
            }
            else
            {
                m_token.erase(0, i);
            }

            if (m_fNegative)
            {
                m_token.insert(0, 1, chNegate);
            }

            return m_token;
        }
    }

    m_token = chZero;

    return m_token;
}

wstring COpndCommand::GetString(uint32_t radix, int32_t precision)
{
    if (m_fInitialized)
    {
        return m_value.ToString(radix, NumberFormat::Float, precision);
    }

    return wstring{};
}

void COpndCommand::Accept(_In_ ISerializeCommandVisitor& commandVisitor)
{
    commandVisitor.Visit(*this);
}

File: src/CalcManager/ExpressionCommand.rs
use std::sync::Arc;

use crate::calc_engine::CCalcEngine;
use crate::calc_utils::Rational;
use crate::calc_display::ICalcDisplay;
use crate::calc_history::CHistoryCollector;
use crate::calc_resource::IResourceProvider;

pub struct CParentheses {
    command: i32,
}

impl CParentheses {
    pub fn new(command: i32) -> Self {
        Self { command }
    }

    pub fn get_command(&self) -> i32 {
        self.command
    }

    pub fn get_command_type(&self) -> CommandType {
        CommandType::Parentheses
    }

    pub fn accept(&self, command_visitor: &mut dyn ISerializeCommandVisitor) {
        command_visitor.visit_parentheses(self);
    }
}

pub struct CUnaryCommand {
    commands: Arc<Vec<i32>>,
}

impl CUnaryCommand {
    pub fn new(command: i32) -> Self {
        Self {
            commands: Arc::new(vec![command]),
        }
    }

    pub fn new_with_two_commands(command1: i32, command2: i32) -> Self {
        Self {
            commands: Arc::new(vec![command1, command2]),
        }
    }

    pub fn get_commands(&self) -> Arc<Vec<i32>> {
        Arc::clone(&self.commands)
    }

    pub fn get_command_type(&self) -> CommandType {
        CommandType::UnaryCommand
    }

    pub fn set_command(&mut self, command: i32) {
        Arc::make_mut(&mut self.commands).clear();
        Arc::make_mut(&mut self.commands).push(command);
    }

    pub fn set_commands(&mut self, command1: i32, command2: i32) {
        Arc::make_mut(&mut self.commands).clear();
        Arc::make_mut(&mut self.commands).push(command1);
        Arc::make_mut(&mut self.commands).push(command2);
    }

    pub fn accept(&self, command_visitor: &mut dyn ISerializeCommandVisitor) {
        command_visitor.visit_unary_command(self);
    }
}

pub struct CBinaryCommand {
    command: i32,
}

impl CBinaryCommand {
    pub fn new(command: i32) -> Self {
        Self { command }
    }

    pub fn set_command(&mut self, command: i32) {
        self.command = command;
    }

    pub fn get_command(&self) -> i32 {
        self.command
    }

    pub fn get_command_type(&self) -> CommandType {
        CommandType::BinaryCommand
    }

    pub fn accept(&self, command_visitor: &mut dyn ISerializeCommandVisitor) {
        command_visitor.visit_binary_command(self);
    }
}

pub struct COpndCommand {
    commands: Arc<Vec<i32>>,
    is_negative: bool,
    is_sci_fmt: bool,
    is_decimal: bool,
    is_initialized: bool,
    token: String,
    value: Rational,
}

impl COpndCommand {
    pub fn new(commands: Arc<Vec<i32>>, is_negative: bool, is_decimal: bool, is_sci_fmt: bool) -> Self {
        Self {
            commands,
            is_negative,
            is_sci_fmt,
            is_decimal,
            is_initialized: false,
            token: String::new(),
            value: Rational::default(),
        }
    }

    pub fn initialize(&mut self, rat: Rational) {
        self.value = rat;
        self.is_initialized = true;
    }

    pub fn get_commands(&self) -> Arc<Vec<i32>> {
        Arc::clone(&self.commands)
    }

    pub fn set_commands(&mut self, commands: Arc<Vec<i32>>) {
        self.commands = commands;
    }

    pub fn append_command(&mut self, command: i32) {
        if self.is_sci_fmt {
            self.clear_all_and_append_command(command);
        } else {
            Arc::make_mut(&mut self.commands).push(command);
        }

        if command == IDC_PNT {
            self.is_decimal = true;
        }
    }

    pub fn toggle_sign(&mut self) {
        for &n_op_code in self.commands.iter() {
            if n_op_code != IDC_0 {
                self.is_negative = !self.is_negative;
                break;
            }
        }
    }

    pub fn remove_from_end(&mut self) {
        if self.is_sci_fmt {
            self.clear_all_and_append_command(IDC_0);
        } else {
            let n_commands = self.commands.len();

            if n_commands == 1 {
                self.clear_all_and_append_command(IDC_0);
            } else {
                let n_op_code = self.commands[n_commands - 1];

                if n_op_code == IDC_PNT {
                    self.is_decimal = false;
                }

                Arc::make_mut(&mut self.commands).pop();
            }
        }
    }

    pub fn is_negative(&self) -> bool {
        self.is_negative
    }

    pub fn is_sci_fmt(&self) -> bool {
        self.is_sci_fmt
    }

    pub fn is_decimal_present(&self) -> bool {
        self.is_decimal
    }

    pub fn get_command_type(&self) -> CommandType {
        CommandType::OperandCommand
    }

    pub fn clear_all_and_append_command(&mut self, command: i32) {
        Arc::make_mut(&mut self.commands).clear();
        Arc::make_mut(&mut self.commands).push(command);
        self.is_sci_fmt = false;
        self.is_negative = false;
        self.is_decimal = false;
    }

    pub fn get_token(&mut self, decimal_symbol: char) -> &str {
        const CH_ZERO: char = '0';
        const CH_EXP: char = 'e';
        const CH_NEGATE: char = '-';
        const CH_PLUS: char = '+';

        self.token.clear();

        for &n_op_code in self.commands.iter() {
            match n_op_code {
                IDC_PNT => self.token.push(decimal_symbol),
                IDC_EXP => {
                    self.token.push(CH_EXP);
                    if self.commands.iter().position(|&x| x == IDC_SIGN).is_none() {
                        self.token.push(CH_PLUS);
                    }
                }
                IDC_SIGN => self.token.push(CH_NEGATE),
                _ => self.token.push_str(&((n_op_code - IDC_0).to_string())),
            }
        }

        for (i, ch) in self.token.chars().enumerate() {
            if ch != CH_ZERO {
                if ch == decimal_symbol {
                    self.token.drain(0..i - 1);
                } else {
                    self.token.drain(0..i);
                }

                if self.is_negative {
                    self.token.insert(0, CH_NEGATE);
                }

                return &self.token;
            }
        }

        self.token = CH_ZERO.to_string();
        &self.token
    }

    pub fn get_string(&self, radix: u32, precision: i32) -> String {
        if self.is_initialized {
            self.value.to_string(radix, NumberFormat::Float, precision)
        } else {
            String::new()
        }
    }

    pub fn accept(&self, command_visitor: &mut dyn ISerializeCommandVisitor) {
        command_visitor.visit_opnd_command(self);
    }
}

pub trait ISerializeCommandVisitor {
    fn visit_opnd_command(&mut self, opnd_cmd: &COpndCommand);
    fn visit_unary_command(&mut self, unary_cmd: &CUnaryCommand);
    fn visit_binary_command(&mut self, binary_cmd: &CBinaryCommand);
    fn visit_parentheses(&mut self, para_cmd: &CParentheses);
}

pub enum CommandType {
    Parentheses,
    UnaryCommand,
    BinaryCommand,
    OperandCommand,
}
